### PR TITLE
Electrostatic boundary condition handling refactor

### PR DESCRIPTION
**Pull request by roelof-groenewald**
This PR follows on #2297, using the new infrastructure created there to also handle the domain boundary conditions / values.

### CODE REVIEW

**Comment by roelof-groenewald** [comment]
@WeiqunZhang I believe I addressed your comment from PR #2297 here.

File: Source/FieldSolver/ElectrostaticSolver.H [hunk 1]
```diff
@@ -32,31 +41,81 @@ struct PhiCalculatorEB {
...
+    std::string potential_zlo_str = "0";
+    std::string potential_zhi_str = "0";
     std::string potential_eb_str = "0";
-    // TODO add handling of domain boundaries as well
+
+    void definePhiBCs ();
 
     void buildParsers ()
     {
+        potential_xlo_parser = makeParser(potential_xlo_str, {"t"});
```
Comment (dpgrote) [inline Source/FieldSolver/ElectrostaticSolver.H]: Dare I ask, but can the potentials on the boundaries also be made spatially dependent? This is a capability I've used in Warp. Though, maybe that should be a separate PR.

Comment (roelof-groenewald) [reply Source/FieldSolver/ElectrostaticSolver.H]: This can be done yes. It would require modifying the `setPhiBC` function in `ElectrostaticSolver.cpp` to properly set the boundary potential. I agree that such a change should probably go in a separate PR.

**Review by RemiLehe — CHANGES_REQUESTED** [review]
Could you add documentation for these new parameters?

**Comment by PhilMiller** [comment]
The 'Intel oneAPI DPC++ SP' CI failure is a compiler crash. It's not a required test, but if we merge it as is, that will likely keep failing on all builds, making it not very useful to run.

I can try to narrow it down, and maybe suggest a workaround in the code until Intel can get it fixed.

File: Source/FieldSolver/ElectrostaticSolver.H [hunk 2]
```diff
@@ -32,31 +41,81 @@ struct PhiCalculatorEB {
...
+        std::set<std::string> eb_symbols = potential_eb_parser.symbols();
+        if ((eb_symbols.count("x") != 0) || (eb_symbols.count("y") != 0)
+                || (eb_symbols.count("z") != 0)) {
+            potential_eb = potential_eb_parser.compile<4>();
+            phi_EB_only_t = false;
+        }
+        else {
+            potential_eb_parser = makeParser(potential_eb_str, {"t"});
+            potential_eb_t = potential_eb_parser.compile<1>();
+        }
```
Comment (PhilMiller) [inline Source/FieldSolver/ElectrostaticSolver.H]: Any reason not to move this function's definition down to the .cpp file?

**Comment by ax3l** [comment]
Note on the DPC++ segfault: We think the latest DPC++ has issues building shared objects (`-fPIC`) and our CI tests DPC++ with Python bindings, which are shared.

First seen in #2279

We can disable either EB or disable the CI entry that builds the DPC++ lib to skip that test, but we need to triage it anyway with @rscohn2 :) In the end, we need to accelerate Python with DPC++, too - thus it's in the CI.

**Comment by ax3l** [comment]
@roelof-groenewald @PhilMiller please rebase. I disabled shared libs for DPC++ temporarily in `development` via #2338

Hope that unblocks you for now until we get this triaged with Intel :)
Update: note https://github.com/ECP-WarpX/WarpX/pull/2279#issuecomment-926852742

If `-fPIC` is not the reason, then let us turn off `EB` for DPC++ builds again.

**Comment by roelof-groenewald** [comment]
@RemiLehe This PR actually has no new parameters added (you already included an entry in the documentation for `warpx.eb_potential(x,y,z,t)` in PR #2310). This PR really just refactors the handling of the domain boundaries to match the way the EB potential is handled. 
Please let me know if you see anything else that should be changed before merging.

**Review by RemiLehe — APPROVED** [review]
